oepcie/yuanphil_impl/oe_ctx_impl.c

```diff
@@ -0,0 +1,221 @@
+		return -1;
+	}
+	ctx -> data.fid = open(ctx -> data.path, O_RDONLY);
+	if(ctx -> data.fid == -1){
+		printf("Failed to open data filestream, returning with error\n");
+		return -1;
+	}
+
+	read(ctx -> header.fid, &(ctx -> config_id), sizeof(int));
+
+	int device_id = 0;
+	int num_devices = 0;
+
+	while(device_id != -1 && ++num_devices){
+		if(device_id > MAX_DEVICE_ID || device_id < 0){
+			printf("Invalid device detected, returning error\n");
+			return -1;
+		}
+		read(ctx -> header.fid, &device_id, sizeof(int));
+		printf("Read device_id %d\n", device_id); // debug
+	}
+	printf("Devices detected: %d\n", num_devices);
+
+	ctx -> map.num_dev = num_devices;
+	ctx -> map.devs = calloc(num_devices, sizeof(device_t));
+
+	// return to start of device_ids
+	lseek(ctx -> header.fid, 4, SEEK_SET);
+
+	for(int i = 0; i < num_devices; i++){
+		int id = read(ctx -> header.fid, &id, sizeof(int));
+		ctx -> map.devs[i] = DEVICES[id];
+	}
+
+	// calculate read offsets
+	for(int i = 1; i < num_devices; i++){
+		ctx -> map.devs[i].read_offset = ctx -> map.devs[i - 1].read_offset +
+			ctx -> map.devs[i - 1].read_size;
+	}
+
+	/*
+	 * Not implemented - Write Offsets... Aren't exactly sure what/if these are
+	 * used yet.
+	 */
+
+	 return 1;
+
+}
+
+int _oe_set_opt(oe_ctx_impl_t* ctx, int option, const char* value){
+	if(ctx -> initialized == 'T'){
+		printf("Already initialized, no more changing options\n");
+		return -1;
+	}
+
+	switch (option){
+		case OE_HEADER_STREAMPATH:
+			ctx -> header.path = realloc(ctx -> header.path, strlen(value) + 1);
+			strcpy(ctx -> header.path, value);
+			break;
+		case OE_CONFIG_STREAMPATH:
+			ctx -> config.path = realloc(ctx -> config.path, strlen(value) + 1);
+			strcpy(ctx -> config.path, value);
+			break;
+		case OE_DATA_STREAMPATH:
+			ctx -> data.path = realloc(ctx -> data.path, strlen(value) + 1);
+			strcpy(ctx -> data.path, value);
+			break;
+		default:
+			printf("Improper option passed, returning with error\n");
+			return -1;
+			break;
+	}
+
+	return 1;
+}
+
+// Allocates a buffer for value, leaves value pointing to a properly formatted
+// dynamic c-string. Returns with error if option is not defined
+int _oe_get_opt(oe_ctx_impl_t* ctx, int option, char* value){
+	switch (option){
+		case OE_HEADER_STREAMPATH:
+			value = malloc(sizeof(char) * strlen(ctx -> header.path) + 1);
+			strcpy(value, ctx -> header.path);
+			break;
+		case OE_CONFIG_STREAMPATH:
+			value = malloc(sizeof(char) * strlen(ctx -> config.path) + 1);
+			strcpy(value, ctx -> config.path);
+			break;
+		case OE_DATA_STREAMPATH:
+			value = malloc(sizeof(char) * strlen(ctx -> data.path) + 1);
+			strcpy(value, ctx -> data.path);
+			break;
+		default:
+			printf("Improper option passed, returning with error\n");
+			return -1;
+			break;
+	}
+	return 1;
+}
+
+// Checks device_id and address for register reads/writes
+int _oe_reg_prep(oe_ctx_impl_t* ctx, int device_id, int addr){
```

Comment:
hehe! When I say "hide" I don't mean hiding the source code. Everything we write here will be perfectly accessible by the user. When I say hide, I'm referring to the separation between _implementation_ and _interface_. The API spec designates the nice shiny library interface that is separated from all the complex and gross stuff going on in the background. It is very important that none of this gross stuff make it past the programming interface because then, in fact, it becomes _part_ of the interface and the user will start writing code with it. Then if we go and change some of the gross stuff, we break their code! Many people confuse this with issues about security or secrecy. That's not the case! We are just trying to make writing application code against our library well defined and consistent.